linker/Linker.Steps/MarkStep.cs

```diff
@@ -1587,5 +1598,116 @@ protected virtual void MarkInterfaceImplementation (TypeDefinition type, Interfa
 			MarkCustomAttributes (iface);
 			MarkType (iface.InterfaceType);
 		}
+
+
+		void MarkSomethingUsedViaReflection (string reflectionMethod, Action<Collection<Instruction>, string, TypeDefinition, BindingFlags> markMethod, Collection<Instruction> instructions)
+		{
+			for (var i = 0; i < instructions.Count; i++) {
+				var instruction = instructions [i];
+				if (instruction.OpCode == OpCodes.Call || instruction.OpCode == OpCodes.Callvirt) {
+					var methodBeingCalled = instruction.Operand as MethodReference;
+					if (methodBeingCalled != null && methodBeingCalled.DeclaringType.Name == "Type" && methodBeingCalled.DeclaringType.Namespace == "System")
+					{
+						if (methodBeingCalled.Name == reflectionMethod) {
+							_context.Tracer.Push ($"Reflection-{methodBeingCalled}");
+							var nameOfThingUsedViaReflection = OperandOfNearestInstructionBefore<string> (i, OpCodes.Ldstr, instructions);
+							var bindingFlags = (BindingFlags) OperandOfNearestInstructionBefore<sbyte> (i, OpCodes.Ldc_I4_S, instructions);
+
+							// There might be more than one ldtoken opcode above the call in the IL stream. Be conservative and check all of
+							// the types which were loaded for the method being used.
+							var declaringTypesOfThingInvokedViaReflection = OperandsOfInstructionsBefore<TypeReference> (i, OpCodes.Ldtoken, instructions);
+							foreach (var declaringTypeOfThingInvokedViaReflection in declaringTypesOfThingInvokedViaReflection) {
+								var typeDefinition = declaringTypeOfThingInvokedViaReflection?.Resolve ();
+								if (typeDefinition != null)
+									markMethod (instructions, nameOfThingUsedViaReflection, typeDefinition, bindingFlags);
+							}
+							_context.Tracer.Pop ();
+						}
+					}
+				}
+			}
+		}
+
+		void MarkConstructorUsedViaReflection (Collection<Instruction> instructions, string unused1, TypeDefinition declaringTypeOfConstructorInvokedViaReflection, BindingFlags bindingFlags)
+		{
+			foreach (var method in declaringTypeOfConstructorInvokedViaReflection.Methods) {
+				if ((bindingFlags == BindingFlags.Default || bindingFlags.HasFlag (BindingFlags.Public) == method.IsPublic) && method.Name == ".ctor")
+					MarkMethod (method);
+			}
+		}
+
+		void MarkMethodUsedViaReflection (Collection<Instruction> instructions, string nameOfMethodInvokedViaReflection, TypeDefinition declaringTypeOfMethodInvokedViaReflection, BindingFlags bindingFlags)
+		{
+			if (nameOfMethodInvokedViaReflection != null) {
```

Comment:
Caller should handle that

Reply:
I would prefer that the caller not handle this null check, because of the constructor case mentioned above.